demo/parser_demo.cpp

```diff
@@ -129,14 +129,61 @@ class PrintVisitor : public ast::Visitor {
             depth--;
         }
     };
-    void visit(ast::Expression* node) override{
-
+    void visit(ast::Expression* node) override {
+        if (node == nullptr) {
+            fmt::print("{:|>{}}- [Expression]> null\n", "", depth);
+        } else {
+            // if binary
+            if (node->operand2) {
+
+                fmt::print("{:|>{}}- [BinaryOperator]> {}\n", "", depth,
+                           node->operation.lit);
+                depth++;
+                node->operand1->accept(*this);
+                node->operand2->accept(*this);
+                depth--;
+            } else {
+
+                fmt::print("{:|>{}}- [UnaryOperator]> {}\n", "", depth,
+                           node->operation.lit);
+                depth++;
+                node->operand1->accept(*this);
+                depth--;
+            }
+        }
     };
-    void visit(ast::UnaryExpression* node) override{
-
+    void visit(ast::Primitive* node) override {
+        if (node == nullptr) {
```

Comment:
[Discussion] Do you think these `if (node == nullptr)` checks in every `visit` function are necessary? I originally added them when I was calling `visit(otherNode)` directly instead of `otherNode->accept(*this)`, which was not the best practice.
They might still be useful in protecting from calling the function directly from outside, but maybe we can check for null there as well?
What do you think?

Reply:
Just from the top of my head: we can replace these raw pointers with `gsl::not_null` from https://github.com/Microsoft/GSL